ui/core/proto_utils/equipped_item.ts

```diff
@@ -219,23 +300,50 @@ export class EquippedItem {
 	}
 
 	withRandomSuffix(randomSuffix: ItemRandomSuffix | null): EquippedItem {
-		return new EquippedItem(this._item, this._enchant, this._gems, randomSuffix, this._reforge);
+		return new EquippedItem({
+			item: this._item,
+			enchant: this._enchant,
+			gems: this._gems,
+			randomSuffix,
+			reforge: this._reforge,
+			upgrade: this._upgrade,
+		});
 	}
 
 	getWithRandomSuffixStats() {
+		return this._getWithDynamicStats();
+	}
+
+	getRandomPropPoints(): number {
+		return this.item.scalingOptions[this._upgrade]?.randPropPoints || this.item.randPropPoints;
+	}
+
+	getWithUpgradeStats() {
+		return this._getWithDynamicStats();
+	}
+
+	_getWithDynamicStats() {
 		const item = this.item;
-		if (this._randomSuffix)
+		if (typeof this._upgrade === 'number') item.stats = item.stats.map((stat, index) => this._item.scalingOptions[this._upgrade]!.stats[index] || stat);
```

Comment:
The `typeof` check will always return `True` and can be deleted.

Reply:
Also, I think you want a `?` rather than a `!` after `this._item.scalingOptions[this._upgrade]`?